### PR TITLE
pyqt6: init

### PR DESCRIPTION
**Pull request by nrdxp**
###### Description of changes

Add a package for pyqt6 which I hope to eventually use to update the webengine backend of Qutebrowser in a follow up PR. 
It is not fully working as #17045 resurfaced and the fix isn't immediately obvious, opening as draft to get some feedback.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by nrdxp** [comment]
@ttuegel, it looks like you fixed this issue in qt5. Any idea how we might be able to do the same here?

**Comment by NickCao** [comment]
That issue seems to be deeply tied with qmake, can we use cmake here as a workaround (it's recommended too)

**Comment by nrdxp** [comment]
Actually I may be able to get around this by crafting a custom build phase and just calling `sip-install`, which is the official build process:
https://www.riverbankcomputing.com/static/Docs/PyQt6/installation.html#building-and-installing-from-source

**Comment by nrdxp** [comment]
nevermind, `sip-build` is still calling qmake under the hood. The switch from qmake to cmake was for qt as a whole itself, but I don't think that extends to individual projects that make use of it. If I really wanted to change the build system to cmake, I'd have to do it in a way that upstream would accept, otherwise it'll just become an unmaintainable hacky mess.

I think the simpler solution is to figure out what `qmake` is missing and fix it.

**Comment by nrdxp** [comment]
@NickCao, I've added a hack commit https://github.com/NixOS/nixpkgs/pull/175510/commits/11882d20adddcdf2c1a000c8c7a6b1595553670d to demonstrate the underlying problem

qtbase is being built without passing the proper flags to the configure phase to point it at the various directories and their locations. The in repo `./configure` script for qtbase has various options such as `-data-dir` to set and install these properly during a build, but we don't use that script in our build of qtbase currently, instead we let qmake handle the configure phase automatically. I' not sure how to pass the equivalent flags to cmake to get it to configure these directories automatically.

In particuar, this line in the qtbase package should not be required:
https://github.com/NixOS/nixpkgs/blob/3688fe70f4ddece54a8a468fa90209066675f0e0/pkgs/development/libraries/qt-6/modules/qtbase.nix#L206-L209

If the various directories are setup correctly before the build, they should be installed in the proper places during the installPhase automatically, so we can remove and try to convince cmake to configure itself properly instead. Maybe we should disable the cmake configure phase and use the `./configure` script instead?

**Comment by NickCao** [comment]
Qt-6 does not support installing everything into different prefixes, or multi output derivations as we call them. Even with the configure script (a cmake wrapper actually), this restriction still cannot be lifted, if I understood correctly.

**Comment by nrdxp** [comment]
Ah okay, that makes sense then. So then maybe the only sane fix is simply not to package the components into different outputs at all? The downside is that this would probably have tree-wide reprocussions. I'm not sure how deep the problem reaches in other parts of nixpkgs, but if they reach far enough, perhaps it would simplify maintenance of all the qt infrastructure, which has thus far seemed to lag behind fairly significantly.

In any case, I think I can at least make the current hack more acceptable for the purposes of this package.

**Comment by NickCao** [comment]
The partial fix supported by upstream (but not sane, from my point of view) is to keep everything inside a single output, but that would greatly bloat the closure size.

**Comment by NickCao** [comment]
And for supporting qmake, we used to have a hacky but working infrastructure in qt-5, fiddling with qmake caches and others, just need the man power to port that all over to qt-6.

**Comment by milahu** [comment]
for qmake see https://github.com/NixOS/nixpkgs/pull/169296#issuecomment-1107636469

i guess #169296 does not work here,
as qmake can use different settings for itself (`dirname("qmake")+"/qt.conf"`)
versus for the project to build (`qmake -qtconf /path/to/qt.conf`)

**Comment by nrdxp** [comment]
> wdym?

This is in reference to the second commit to this PR, which I would like to remove with a cleaner fix. Basically I just combine the `qtbase` `dev` and `out` attributes into a single derivation to please qmake (so that it could find the architecture specifications). 

I'd rather do this with a qt.conf. The docs for `QMAKESPEC` made it seem like I had to point it at a specific architecture specific configuration, meaning I would have to create a complex nix function to translate support Nix arches to their qmake equivalent names. Perhaps I just misundertstood this though, I'll try playing with it.

**Comment by milahu** [comment]
> combine the `qtbase` `dev` and `out` attributes into a single derivation to please qmake

yepp, qt doesn't like nix's multi-prefix filesystem

instead of fixing qmake, you could try porting [pyqtbuild](https://www.riverbankcomputing.com/hg/PyQt-builder/file/tip/pyqtbuild/builder.py) to cmake with [pro2cmake](https://code.qt.io/cgit/qt/qtbase.git/tree/util/cmake/pro2cmake.py) (aka qmake2cmake)
pyqtbuild uses `qmake -query` to build `self.qt_configuration`
then generates a `.pro` file and calls `self._run_qmake`

**Comment by milahu** [comment]
> you could try porting [pyqtbuild](https://www.riverbankcomputing.com/hg/PyQt-builder/file/tip/pyqtbuild/builder.py) to cmake

[i gave it a try](https://github.com/NixOS/nixpkgs/commit/a19e174e88edd5ae2359e90cdcd3470f14f1cc6d). im stuck at
`AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'`

```
  [ 93%] Built target QtWebChannel
  [ 95%] Automatic MOC and UIC for target QtWebSockets
  [ 95%] Built target QtWebSockets_autogen
  Consolidate compiler generated dependencies of target QtWebSockets
  [100%] Built target QtWebSockets
  Install the project...
  -- Install configuration: ""
  -- Installing: /build/tmp82gabk4e/lib64/libQtNetwork.so
  -- Set runtime path of "/build/tmp82gabk4e/lib64/libQtNetwork.so" to "$ORIGIN:$ORIGIN/../lib64"

[...]

  -- Installing: /build/tmp82gabk4e/lib64/libQtWebChannel.so
  -- Set runtime path of "/build/tmp82gabk4e/lib64/libQtWebChannel.so" to "$ORIGIN:$ORIGIN/../lib64"
  -- Installing: /build/tmp82gabk4e/lib64/libQtWebSockets.so
  -- Set runtime path of "/build/tmp82gabk4e/lib64/libQtWebSockets.so" to "$ORIGIN:$ORIGIN/../lib64"
  Traceback (most recent call last):
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      hook = backend.prepare_metadata_for_build_wheel
  AttributeError: module 'sipbuild.api' has no attribute 'prepare_metadata_for_build_wheel'

  During handling of the above exception, another exception occurred:

  Traceback (most recent call last):
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      main()
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      json_out['return_val'] = hook(**hook_input['kwargs'])
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      return _get_wheel_metadata_from_wheel(whl_basename, metadata_directory,
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      dist_info = _dist_info_files(zipf)
    File "/nix/store/d5ll8w6x4b75rid43s6hfwf7hlxcsqbk-python3.9-pip-22.0.4/lib/python3.9/site-packages/pip/_vendor/pep517/in_process/_in_process.p>
      raise Exception("No .dist-info folder found in wheel")
  Exception: No .dist-info folder found in wheel
  error: subprocess-exited-with-error
```

**Comment by nrdxp** [comment]
> your `sip-build` solution looks better

unfortunately it still relies on the combined derivation hack, but if we could finish porting pyqtbuild to cmake, I think that would do it. I've at least got the package to build. Just need to find a reasonable solution for the hack now.

**Comment by milahu** [comment]
to make the build faster:

```py
# PyQt-builder-1.12.2/pyqtbuild/builder.py
    def get_options(self):

        default_jobs = os.environ.get("NIX_BUILD_CORES", 1)
        options.append(
                Option('jobs', option_type=int,
                        default=default_jobs,
                        help="run N make jobs in parallel", metavar='N'))
```

the first phase of the build (codegen) is still slow (single core)

setting qt paths via qt.conf (https://github.com/NixOS/nixpkgs/pull/174306#issuecomment-1148842345) looks nice
but qmake cannot find qtdeclarative, and checkPhase throws

```
ModuleNotFoundError: No module named 'PyQt6.QtQml'
```

qmake vs nix multi prefix ...

File: pkgs/development/python-modules/pyqt/6.x.nix [hunk 1]
```diff
@@ -0,0 +1,158 @@
...
+
+  configurePhase = ''
+    runHook preConfigure
+
+    export PYTHONPATH=$PYTHONPATH:${out}/lib/${python.libPrefix}/site-packages
+
+    sip-build --no-make \
+     --confirm-license \
+     --build-dir=dist \
+     --target-dir=${out}/lib/${python.libPrefix}/site-packages \
```
Comment (milahu) [inline pkgs/development/python-modules/pyqt/6.x.nix:110]: `lib/${python.libPrefix}/site-packages` == `${python.sitePackages}`

**Comment by peterhoeg** [comment]
> Basically I just combine the `qtbase` `dev` and `out` attributes into a single derivation to please qmake (so that it could find the architecture specifications).

We have similar issue with LibreOffice when building with KDE support where the builder assumes that all components are under the same prefix - and the *fix* was to use `symlinkJoin` to give it all the files it needs.

It's not pretty and really rubs me the wrong way, but it's preferred to having to deal with heavy patching of the build scripts.

```nix
{
  # See `postPatch` for details
  kdeDeps = symlinkJoin {
    name = "libreoffice-kde-dependencies-${version}";
    paths = flatten (map (e: [ (getDev e) (getLib e) ]) [
      qtbase
      qtx11extras
      kconfig
      kcoreaddons
      ki18n
      kio
      kwindowsystem
    ]);
  };
}
```

**Comment by nrdxp** [comment]
> and the _fix_ was to use `symlinkJoin` to give it all the files it needs

Yeah I hear you, unfortunately something about qmake didn't like symlinks. This was the first approach I took but it unfortunately didn't work out. I am assuming qmake was just programmed to expect regular files. :frowning_face:

**Comment by milahu** [comment]
lets just merge this with the workaround eeac59597794ba90046847dae895aa2197beb37e

**Pull request closed without merging**